### ISSUE TITLE
painful to get started

### ISSUE THREAD

**Issue by boneskull** [rollup/rollup#385]
This project needs some better tutorials or examples.  Here's my use case:
1.  I have code written in ES2015; I am using Babel to transpile
2.  I am also using npm dependencies
3.  These dependencies (mainly) use CommonJS instead of ES2015 modules
4.  This is a test runner with a programmatic API
5.  The resulting build(s) must be runnable in Node.js, a browser, and a CLI context

I've discovered the Babel plugin, the NPM plugin, and the CommonJS plugin.
It's unclear:
1.  In what order the plugins are supposed to run
2.  If order actually matters (though I'm pretty sure it does)
3.  Why the CommonJS plugin and NPM plugin are separated at all
4.  How to configure the Babel plugin to work with the above two plugins
5.  Why the Babel plugin is necessary
6.  Where Babel's responsibilities end and Rollup's begin--don't you need Babel to use `import` statements anyway?
7.  If/when Browserify or Webpack is necessary
8.  How to configure Karma to make my tests run in a browser
9.  How to explain to my users, in a README, how to include my library given a matrix of potential configurations (Node.js, browser, Babel, no-Babel, Rollup, etc)
10. When to use a config file or when and how to use a build script (`import rollup from 'rollup'`) 

I'm here asking these questions, there's probably 5 other people who had the same ones and didn't post an issue.  Surely I'd like my questions addressed, but the real need is more documentation.

Thanks!

**Comment by boneskull** [comment]
As an aside, Babel "presets" are very helpful in that they provide an abstraction which doesn't require the user to manually include *n* modules and configure them all.  Perhaps packages for common use cases would take some responsibility off of the user's shoulders?

**Comment by eventualbuddha** [comment]
@boneskull Thanks for the thorough issue notes! I assume you've looked at the [wiki](https://github.com/rollup/rollup/wiki)? Were you looking for a Getting Started page and didn't find it? Did you find [rollup-starter-project](https://github.com/rollup/rollup-starter-project)? Here are a few answers to your questions, though as you said the better thing to do is to create better documentation.

> In what order the plugins are supposed to run
> If order actually matters (though I'm pretty sure it does)

The order does matter since they run in the order listed in the `plugins` array for each of their [hooks](https://github.com/rollup/rollup/wiki/Plugins#creating-plugins). You can think of plugins as a pipeline, where each plugin can modify the results of the one before it.

> How to configure the Babel plugin to work with the above two plugins?

There shouldn't be any special configuration needed, for the most part. Are you seeing problems that led to this question?

> Why the CommonJS plugin and NPM plugin are separated at all

We've [talked about](https://github.com/rollup/rollup/issues/360#issuecomment-165516896) making a `rollup-plugin-node` package that combines `rollup-plugin-commonjs` and `rollup-plugin-npm`, as well as adding support for other things like `__dirname`.

> Why the Babel plugin is necessary?
> Where Babel's responsibilities end and Rollup's begin--don't you need Babel to use `import` statements anyway?

The Babel plugin exists to transpile ES2015 code to ES5 code, but leaves the `import` and `export` statements intact because rollup handles them itself.

> When to use a config file or when and how to use a build script (`import rollup from 'rollup'`)

The canonical way to do this these days it to make one (or many) `rollup.config.js` file that imports plugins etc and exports a configuration object. You can see [`rollup.config.js`](https://github.com/rollup/rollup/blob/master/rollup.config.js) from rollup itself for an example.

There are other things to discuss, but the important thing is to figure out where these things should be documented. Thanks for spurring the discussion!

**Comment by Rich-Harris** [comment]
@boneskull this is awesome feedback, thanks! Have been thinking about producing some screencasts to help people get started – if I get round to it, these questions would provide a great starting point.

**Comment by thisconnect** [comment]
Hi, first of all I absolutely love rollup's (and babels) generated code. I'd like to add some beginners questions as well, this might turn in some kind of FAQ :)

-  some examples in the wiki, use ES2015 syntax for the rollup JavaScript API ( [Plugins#using-plugins](https://github.com/rollup/rollup/wiki/Plugins#using-plugins) and [JavaScript-API#plugins](https://github.com/rollup/rollup/wiki/JavaScript-API#plugins) ). How do you run a build script (JavaScript API) written in ES2015? Do you use babel-cli somehow?

- To split code into 2 files is it recommended to use 2 configs using exclude and different entry points?

**Comment by rosskevin** [comment]
I'm new to rollup and working to integrate es6 to the v4 of bootstrap-material-design.  As a new user, hear's my experience:

1. what the h*** do I need to use? babel 6 was working great with chrome, but now that I need Symbol.iterator polyfill's for firefox and can't get them to work, looks like I need something to streamline my `babel -> concat -> stamp -> uglify` into a package that doesn't ignore my `import babel-polyfill`
1. most of our users are non-node, but we have node, meteor, browser only, rails, etc
1. a lot of our users use bower, how does that impact this choice?
1. browserify?  I think I just want a simple package for for module users and non-module users since this manifests as a jquery plugin
1. jspm seems too much for this, I just want a build time tool to put this thing together in various module formats
1. oh, so rollup might be good for a library based on the readme's

    > so is a better choice for creating libraries
1. hmm, I'm comfortable with babel and have a working config, what do I need to do here...
1. ... long stretch of confusion ...
1. ah, found the starter project, cool. I'm naming my local files `rollup.*` to give some indication what it is from.
1. hmm, this shows umd and es6...but some of the plugins generate cjs.  Which packages are best practice to produce for a library?  I guess umd covers cjs, any downsides to that?
1. grunt...hmm...this project uses grunt but I like the starter config files, I guess I'll just use exec

 ```
    exec: {
      'rollup-umd': {
        command: 'rollup -c rollup.umd.config.js'
      },
      'rollup-es6': {
        command: 'rollup -c rollup.es6.config.js'
      }
    }
```
- yay, I have my files! Ouch, uh, no sourcemaps.

That's what I've gotten so far.  Hopefully it will give an idea what I encountered.  You might put something at the top of the readme like:

> TLDR;   See rollup-starter-project for inspiration on how to get started.  UMD, and ES6 examples with Babel included.

**Comment by thisconnect** [comment]
@rosskevin about 11. just use npm scripts
https://docs.npmjs.com/cli/run-script
https://docs.npmjs.com/misc/scripts

add 
```json
{
  "scripts": {
    "build-umd": "rollup -c rollup.umd.config.js"
  }
}
```

and run `npm run build-umd`

**Comment by Rich-Harris** [comment]
Have made a start on a guide at http://rollupjs.org/guide/. It's highly incomplete, but feedback is welcome nonetheless, particularly if there are specific areas you think we should cover.

**Comment by rosskevin** [comment]
@Rich-Harris great start! That would have answered most of my questions.  

Still remaining for me that I think should be in the guide:
1. sourcemaps - adding `sourceMap: true,` to `.babelrc` has no output
2. self executing bundle - would like to see an example.  The lib I'm working on will need the simple demodularized version as it manifests as a jquery plugin.

**Comment by Victorystick** [comment]
> * 1. sourcemaps - adding `sourceMap: true,` to `.babelrc` has no output

It should be added to the `bundle.generate` call or to your Rollup configuration, not `.babelrc`.
```diff
...

export default {
  entry: 'main.js',
+ sourceMap: true,
  plugins: [ ... ]
};
```

If you're using the CLI you'll likely want `rollup --sourcemap inline`. See `rollup -h` for all options.

> * 2. self executing bundle - would like to see an example

All bundles are "self executing", see [this example that uses jQuery](http://rollupjs.org/#%7B%22options%22%3A%7B%22format%22%3A%22iife%22%2C%22moduleName%22%3A%22myBundle%22%2C%22globals%22%3A%7B%22jquery%22%3A%22jQuery%22%7D%7D%2C%22modules%22%3A%5B%7B%22name%22%3A%22main.js%22%2C%22code%22%3A%22%2F%2F%20You%20can%20import%20external%20modules%20into%20your%20bundle%20%E2%80%93%5Cn%2F%2F%20it%20doesn't%20matter%20if%20they're%20ES2015%20or%20legacy%5Cnimport%20%24%20from%20'jquery'%3B%5Cn%5Cn%24(%20'body'%20).html(%20'%3Ch1%3EHello%20world!%3C%2Fh1%3E'%20)%3B%22%7D%5D%7D) to do some simple DOM manipulation. If you're adding a jQuery method, you should be able to do something like this.

```js
import { fn } from 'jquery';

fn.extend({
  myMethod: ...
});
```

**Comment by rosskevin** [comment]
@Victorystick regarding self executing - perhaps I'm not using the proper nomenclature - I want old school users to be able to include the js without a module manager.  Perhaps this is the `iife` format?

**Comment by Victorystick** [comment]
@rosskevin Indeed it is. :smile:

**Comment by rosskevin** [comment]
I can start a new issue at any time, it seems like my experience as a new user might be useful for this discussion.  Just let me know.

**Importing legacy scripts*** - for my project's documentation js, I need several `vendor` scripts that I'm planning to rollup, I'm expecting that a simple `import` will do the trick e.g.

```
import anchors from '../vendor/anchor.min.js'
import Clipboard from '../vendor/clipboard.min.js'
import holder from '../vendor/holder.min.js' 
````
When I execute rollup, I receive `vendor/anchor.min.js does not export default `.  Should I be `import`ing based on the file?  Previously I used babel and uglify, and the line looked like `import anchors from 'anchor'`, but this would give rollup zero information about the file location which I assume would mean that it would not be bundled.  Perhaps this is a case where I need to stick with uglify for concatenation?

**Comment by Victorystick** [comment]
> **Importing legacy scripts**

I assume these vendored scripts are not written as ES modules? Rollup can only processes modules written using `import`/`export` syntax, but you can use the [`commonjs-plugin`](https://github.com/rollup/rollup-plugin-commonjs) to support such modules if that's what you need. See #45

**Comment by rosskevin** [comment]
I opted for the uglify of vendored scripts, it makes sense for this situation.

The very last issue I have to graduate from rollup newbie to user, which actually led me down this long path, is to include the polyfill for firefox.  Specifically, I was originally [trying to `import 'babel-polyfill'`](https://babeljs.io/docs/usage/polyfill/) in order to avoid firefox `ReferenceError: Symbol is not defined`.  When I try to `import 'babel-polyfill'` I receive `ReferenceError: babelPolyfill is not defined`.  Do I need to import it from it's `node_modules` location? Or is there a preferred way to get polyfills with rollup?

**Comment by rosskevin** [comment]
So to get my rollup package created for the browser (not node based), I needed to import the polyfill via `import '../../node_modules/babel-polyfill/dist/polyfill'`.  

My recommendation for the guide is to give a short discussion on bundling external modules and excluding them.  Something that might expand on a comment I found in a sample file:

```
/*
 * This is the main entry point for your package.
 *
 * You can import other modules here, including external packages. When
 * bundling using rollup you can mark those modules as external and have them
 * excluded or, if they have a jsnext:main entry in their package.json (like
 * this package does), let rollup bundle them into your dist file.
 */
```

I've got everything packaging now with rollup, and with significantly less grunt config code.

**Comment by sethwklein** [comment]
My goal was simple: make a browser bundle including React and PouchDB.

* I've used Browserify and webpack.
* I've read and understood the code they produce.
* I found the Rollup getting started guide, the README, the wiki, and a bunch of issues.
* I did a ton of Google searching.

It took me 7 hours.

While the Rollup guide might be a starting point for library authors, it lacks several things required for creating real world browser bundles, including a couple that stand out as worthy of additional mention:

* Dependency resolution. People coming from Node/Browserify/webpack don't know there is any way but the Node way and are unsure why anyone wouldn't "just" do it that way by default.
* CommonJS dependencies. These are common (heh) and while `jsnext:main` is out there, it doesn't always even work. (PouchDB is a case in point.)
* Just because a package uses Node globals, doesn't mean it's intended for the browser.
* Output format. When things are working, IIFE is an obvious choice (if you even know what that acronym stands for) but when they're not, when none of the formats produce a working bundle, it becomes much more difficult to figure out which choice you need.

Rollup needs a guide for making real world browser bundles. The above are some of the points it needs to hit.

In the meantime, perhaps the following will save 5 or 6 hours per person who tries to get it working:

```js
// works when running rollup with node 4.2.1
import nodeResolve from "rollup-plugin-node-resolve"
import commonjs from "rollup-plugin-commonjs"
import replace from "rollup-plugin-replace"
import globals from "rollup-plugin-node-globals"
import json from "rollup-plugin-json"
import babel from "rollup-plugin-babel"
// import uglify from "rollup-plugin-uglify"

export default {
  entry: "main.jsx",
  format: "iife",
  dest: "bundle.js",
  plugins: [
    nodeResolve({
      browser: true,
      preferBuiltins: false,
    }),
    commonjs({
      ignoreGlobal: true,
    }),
    globals(),
    json(),
    babel({
      exclude: "node_modules/**",
      // plugins appears to be ignored. use .babelrc
    }),
    // uglify(),
  ],
}
```

**Comment by hollowdoor** [comment]
One question I have is -- How do I bundle in a test folder? This one library I'm working on requires testing in a browser.

I could just have the rollup commonjs bundle, and browserify it. Or use SystemJS. It would rather still be nice to see some instruction on how to manage the testing process with rollup. Whether, or not a test framework is involved.

At best my idea right now is to add some package scripts like:

```json
{
    "scripts": {
        "roll": "rollup -c",
        "browser": "browserify test/src.js > test/code.js",
        "build": "npm run roll && npm run browser"
    }
}
```

@sethwklein Your solution for "real world bundles" appears related. There is some cross over in the two problems.

I haven't been able to find any documentation for how to manage multiple sources.

Maybe place another rollup config in the test folder and:

```json
{
    "scripts": {
        "roll": "rollup -c",
        "browser": "cd test && rollup -c",
        "build": "npm run roll && npm run browser"
    }
}
```

**Comment by rarkins** [comment]
So is there any "simple" guide yet for bundling JS for the browser? Taking into account that most dependencies use CommonJS today of course.

**Comment by snuggs** [comment]
@rarkins care to provide a use case? There are various _"micro-strategies"_ that exist that can run in concert. I currently use rollup/buble in tandem and takes care of 100% of my browser concerns. My colleague @btakita probably has better insight as we are using rollup in production. /cc @Victorystick @TrySound

**Comment by rarkins** [comment]
@snuggs I currently use browserify in one instance and webpack in another. In both cases it's browser JS  using a lot of CommonJS "require()" statements to pull in both local code (relative paths) as well as node_modules (e.g. `var keycode = require('keycode');`).

The webpack case involves a lot of angular modules. As per "best practices" we've split modules up a lot, the result being many relative and nested `require()` statements. After reading lately about rollup's flattening of requires (as opposed to webpack/browserify which embed their own require statements) and the efficiency of doing so, I'm interested in trying rollup to build these same "bundles" and see if they're more efficient to execute.

I'm using HTTP2 in production with server push, so have decided to "unbundle" vendor files from my client code and hence I'm not after any tree shaking.

So in summary I have about a dozen "entry point" JS files that have a series of nested require statements + code and I want those bundled up, like I can do with browserify and webpack.

**Comment by snuggs** [comment]
Considered rollup-multientry plugin? 

I'm currently using that and rollem in the meantime:
https://github.com/bahmutov/rollem
https://github.com/rollup/rollup-plugin-multi-entry

Let me know if you need any setup help.

**Comment by lukekentwell** [comment]
I am having trouble configuring globals in a config file. 

I attempted the following (following on from the CLI format)

`globals: "d3-selection:d3,d3-collection:d3,d3-dispatch:d3,d3-scale:d3,d3-array:d3",`

and (following on from the javascript API format)

`globals: {
        "d3-selection": "d3",
        "d3-collection": "d3",
        "d3-dispatch": "d3",
        "d3-scale": "d3",
        "d3-array": "d3"
    },`

However neither worked, I'm sure I'm missing something simple, can anyone shed some light on how to do this?

**Comment by hollowdoor** [comment]
Where should a person contribute to improve the docs? The wiki?

**Comment by thisconnect** [comment]
> Where should a person contribute to improve the docs? The wiki?

Yes wiki or https://github.com/rollup/rollupjs.org/tree/master/src/guide

**Comment by hollowdoor** [comment]
Thanks @thisconnect. How should we decide where to contribute? For instance I would like to address some of the problems brought up here. With a ton of searching, and trial, and error I've figured out a few things. :)

Also there are solutions here that would probably work for docs.

**Comment by mbonaci** [comment]
Regarding migration from Webpack, has anyone actually suggested making rollup parse webpack config and output rollup config? Or, to avoid polluting code base, have a site where you could paste WP config and get rollup (at least some starter config with comments where migration is impossible/unsuccessful).
People would be able to quickly see the difference between the two builds.

**Comment by jhabdas** [comment]
@sethwklein I blame Node. And here is [how they shall pay](https://github.com/jhabdas/fetch-inject).

**Comment by hollowdoor** [comment]
@jhabdas How is node to blame?

**Comment by jhabdas** [comment]
@hollowdoor That was a joke. But I still decided to remove "main" from my package manifest in favor of "module" to help push the envelope on ES6 Modules out of empathy for Web developers.

**Comment by jhabdas** [comment]
Nope. Still in draft spec dated Jan '16 AFAICT

**Comment by hollowdoor** [comment]
I see. Anyway. `fetch-inject` looks interesting. I might give it a try some time.

Something just occurred to me. These bundlers like browserify, and rollup all require names to be specified in the config for exporting global variables. I feel that why that is hasn't been sufficiently explained. There's file names, and the `package.json` name field to get a name. I suppose it has to do with global naming conflicts in a webpage across script tags. I would like to find out if there's a way to automatically resolve a module name to produce a global. So far I haven't seen one in rollup.

**Comment by alwaysonlinetxm** [comment]
@sethwklein
I used rollup-plugin-node-globals and rollup-plugin-babel together in my project. But I found that they can't work together.

I still got these warnings below as I have used  rollup-plugin-node-globals:

```JavaScript
$ NODE_ENV=development rollup -c -w 
bundling...
⚠️   'url' is imported by node_modules/node-fetch/index.js, but could not be resolved – treating it as an external dependency
https://github.com/rollup/rollup/wiki/Troubleshooting#treating-module-as-external-dependency

⚠️   'http' is imported by node_modules/node-fetch/index.js, but could not be resolved – treating it as an external dependency
https://github.com/rollup/rollup/wiki/Troubleshooting#treating-module-as-external-dependency

⚠️   'https' is imported by node_modules/node-fetch/index.js, but could not be resolved – treating it as an external dependency
https://github.com/rollup/rollup/wiki/Troubleshooting#treating-module-as-external-dependency

⚠️   'zlib' is imported by node_modules/node-fetch/index.js, but could not be resolved – treating it as an external dependency
https://github.com/rollup/rollup/wiki/Troubleshooting#treating-module-as-external-dependency

⚠️   'stream' is imported by node_modules/node-fetch/index.js, but could not be resolved – treating it as an external dependency
https://github.com/rollup/rollup/wiki/Troubleshooting#treating-module-as-external-dependency

...
```

actually rollup-plugin-babel dosen't work well too. And there is my project to reproduce this problem:
https://github.com/alwaysonlinetxm/rollup_issue.git
(Just run yarn install & yarn run start and you will see some warnings)

**Comment by jhabdas** [comment]
@alwaysonlinetxm and what happened when you followed the instructions in the troubleshooting doc output 5 times in that output at least 5 times?

**Comment by alwaysonlinetxm** [comment]
@jhabdas Sorry.. I don't get what you mean..

**Comment by jhabdas** [comment]
> @jhabdas Sorry.. I don't get what you mean..

Please see this link for resolution: https://github.com/rollup/rollup/wiki/Troubleshooting#treating-module-as-external-dependency

**Comment by alwaysonlinetxm** [comment]
@jhabdas I knew that, but it's not my problem..And I have found what happened, it's my own mistake. Thank you all the same!

**Comment by jhabdas** [comment]
@alwaysonlinetxm no problem. we all make mistakes. if possible, please try and leverage them as opportunities to improve the docs.

**Comment by Announcement** [comment]
@boneskull how's the problem what do you still need figured out? do a quick recap please

**Comment by boneskull** [comment]
I haven't really touched rollup since I posted this issue.

**Comment by hollowdoor** [comment]
@boneskull In my opinion rollup is the best bundler. The only problem is the docs. Especially plugin docs which don't tend to emphasize the order in which they should be put in rollup, and some other gotchas. I've gotten passed that on my own so I don't really worry about it now, but there's still people new to rollup.

To everybody else: I think what rollup needs is recipes, or what might be called tutorials for common setups. There are some out there in some blogs, but rollup could benefit from some official recipes.

Another option could be compound plugins. The normal plugins could stay as is, and then there could be compound plugins which depend on, and pull in other plugins. rollup's internals would have to be changed to support this probably. For instance a compound "babel/react/commonjs/node resolve" plugin would reduce a lot of strife for many new users of rollup.

For official compound plugins a survey could be done to collect info about the majority of plugin combinations that people use, and the devs of rollup could implement just those as official plugins. Letting odd compound plugins get developed by third parties.

I'm not too familiar with rollup internals so I probably wouldn't do the PR for compound plugins. I'm just an idea man at this point.

**Comment by jhabdas** [comment]
> I think what rollup needs is recipes.

Collecting this information in a central location seems like a great idea. Perhaps someone could update the Wiki or start dumping information here. Any takers?

**Comment by hollowdoor** [comment]
An attempt.

`buble` cross node/browser config
-------------------------

This describes a basic `rollup.config.js` for building a module that can work in browsers, and node.js. Some information is left out for brevity. See [using config files](https://rollupjs.org/#using-config-files).

See these plugins to learn more:

* [node-resolve](https://github.com/rollup/rollup-plugin-node-resolve)
* [commonjs](https://github.com/rollup/rollup-plugin-commonjs)
* [buble](https://gitlab.com/Rich-Harris/rollup-plugin-buble#README)
* [buble website](https://buble.surge.sh/)

Make a `rollup.config.js` in the root of your project.

```javascript
// rollup.config.js
import nodeResolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import buble from 'rollup-plugin-buble';
const external = Object.keys(require('./package.json').dependencies || {});

export default {
  input: 'src/main.js',
  external: external,
  output: {
    file: 'dist/bundle.js',
    format: 'cjs' //So the module is acceptable to the node.js environment
  },
  plugins: [
      //Plugins should be in this order
      //1. nodeResolve
      //2. commonjs
      //3. buble
      nodeResolve({
          jsnext: true,
          main: true,
          module: true
      }),
      commonjs(),
      buble()
  ]
};
```

In your `package.json`:

```json
{
    "name": "happy-face",
    "main": "dist/bundle.js",
    "jsnext:main": "src/main.js",
    "module": "src/main.js"
}
```

In `src/main.js`:

```javascript
const smiley = ':)';
export function happyFace(){
    return smiley;
}
```

After creating the above files run `rollup -c` in the directory that contains `rollup.config.js`.

The `"jsnext:main"`, and `"module"` fields point to an un-transpiled version of ECMAscript 2016 (ES6).

**Hint:** Swap `babel` for `buble` in the above `rollup.config.js` to use the `babel` rollup plugin instead (Though this will take some configuration of babel).

No node natives will get transpiled while using this recipe.

After doing all of the above your module should be ready to be published to [npm](https://www.npmjs.com/). Others will be able to use your module in the browser too (if they are also using babel, rollup, or some such thing).

**Comment by hollowdoor** [comment]
A gulp/rollup/babel tutorial I found https://medium.com/@andrewhenderson/es6-with-babel-6-gulp-and-rollup-aa7aeddeccc6

**Comment by guybedford** [comment]
Closing as this has been inactive for a while and newer relevant issues can track these concerns.
